src/System.Private.ServiceModel/src/System/ServiceModel/Channels/StreamSecurityUpgradeInitiatorBase.cs

```diff
@@ -99,33 +75,48 @@ public override Stream InitiateUpgrade(Stream stream)
             return result;
         }
 
-        internal override void EndOpen(IAsyncResult result)
+        internal override async Task<Stream> InitiateUpgradeAsync(Stream stream)
         {
-            base.EndOpen(result);
-            _isOpen = true;
+            if (stream == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("stream");
+            }
+
+            if (!_isOpen)
+            {
+                 Open(TimeSpan.Zero);
+            }
+
+            var remoteSecurityWrapper = new OutWrapper<SecurityMessageProperty>();
+            Stream result = await OnInitiateUpgradeAsync(stream, remoteSecurityWrapper);
+            _remoteSecurity = remoteSecurityWrapper;
+            _securityUpgraded = true;
+            return result;
         }
 
         internal override void Open(TimeSpan timeout)
         {
-            base.Open(timeout);
             _isOpen = true;
         }
 
-        internal override void EndClose(IAsyncResult result)
+        internal override Task OpenAsync(TimeSpan timeout)
         {
-            base.EndClose(result);
-            _isOpen = false;
+            _isOpen = true;
+            return Task.CompletedTask;
         }
 
         internal override void Close(TimeSpan timeout)
         {
-            base.Close(timeout);
             _isOpen = false;
         }
 
-        protected abstract IAsyncResult OnBeginInitiateUpgrade(Stream stream, AsyncCallback callback, object state);
-        protected abstract Stream OnEndInitiateUpgrade(IAsyncResult result,
-            out SecurityMessageProperty remoteSecurity);
+        internal override Task CloseAsync(TimeSpan timeout)
+        {
+            _isOpen = false;
```

Comment:
Same as for OpenAsync, call the synchronous Close method so logic in one place. Compiler will inline.